src/lib/updateGrantsGrantees.js

```diff
@@ -0,0 +1,115 @@
+import AdmZip from 'adm-zip';
+import { toJson } from 'xml2json';
+import {} from 'dotenv/config';
+import axios from 'axios';
+import {
+  Grantee, Grant,
+} from '../models';
+import logger from '../logger';
+
+const fs = require('mz/fs');
+/**
+ * Reads HSES data files that were previously extracted to the "temp" directory.
+ * The files received from HSES are:
+ *
+ * agency.xml - grantee and grantee that are delegates
+ * grant_agency.xml - junction between grants and agencies
+ * grant_award.xml - grants
+ * grant_award_replacement.xml
+ * grant_program.xml
+ *
+ * The grantee data is them filtered to exclude delegates
+ *
+ */
+async function processFiles() {
+  let grantGrantees;
+  let grants;
+  const granteesForDb = [];
+  const grantsForDb = [];
+
+  try {
+    const grantAgencyData = await fs.readFile('./temp/grant_agency.xml');
+    const json = toJson(grantAgencyData);
+    const grantAgency = JSON.parse(json);
+    // we are only interested in non-delegates
+    grantGrantees = grantAgency.grant_agencies.grant_agency.filter(
+      (g) => g.grant_agency_number === '0',
+    );
+
+    // process grantees aka agencies that are non-delegates
+    const agencyData = await fs.readFile('./temp/agency.xml');
+    const agency = JSON.parse(toJson(agencyData));
+
+    // filter out delegates by matching to the non-delegates
+    // eslint-disable-next-line max-len
+    const grantees = agency.agencies.agency.filter((a) => grantGrantees.some((gg) => gg.agency_id === a.agency_id));
+
+    grantees.forEach((g) => granteesForDb.push({
+      id: parseInt(g.agency_id, 10),
```

Comment:
I think we need a new `externalId` or `hsesId` column for both `Grants` and `Grantees` to use for the IDs we get from this data, and allow the `id` column to be auto-incremented by postgresql. Explicitly assigning to the auto-incremented column will cause problems down the road.

How are we merging (or matching) the grantees from this data source and those pulled in via the tta goals?

Reply:
I could also see removing the auto-increment and relying on HSES to provide the IDs for these two classes, though I'm a little squeamish on trusting any canonical IDs in our system to any outside system.

Reply:
> How are we merging (or matching) the grantees from this data source and those pulled in via the tta goals?

That is on my list of modifications to the goals import script. The main idea was to change the code to find grants and grantees in the database and use the ids for the GrantGoals table - if running in production or other deployed environments. For local development, I'll have to think what would be the best solution.

Reply:
I could also see removing the auto-increment and relying on HSES to provide the IDs for these two classes, though I'm a little squeamish on trusting any canonical IDs in our system to any outside system.

If we could re-use HSES ids which are already present in 3 tables, it would give us "free" ids. I was envisioning that all grants and grantees would come from HSES, so we wouldn't need to insert any real grantees and grants. If that's a safe assumption, then we could adjust how we are entering test data.

The alternative would be to rely on auto increment for the HSES pull, but then we couldn't use the bulk operations and rely on something like "findOrCreate". This is fine, but not as fast as the bulk inserts and updates. The only advantage of the autoincrement would be an easier handling of the test data (unless I am missing something)

Reply:
> I was envisioning that all grants and grantees would come from HSES, so we wouldn't need to insert any real grantees and grants

this is almost certainly true.

My biggest worry is _if_ we find out that a grantee for whatever reason gets assigned a new ID in hses, how do we handle it on our side. My impression had been that there would be a new agency object on each new 5 year grant cycle. But I think we've seen in the data that that would not be the case. There would be one agency with the old inactive grant and a new active grant, right?

> For local development, I'll have to think what would be the best solution.

if we go the explicit ID route, we could use the test data in /temp/*.xml as our test/dev data across the whole system. Add them to seeders like the users and permissions used in those environments.